File: blue-core/src/blue/soundObject/AudioFile.java
/*
 * blue - object composition environment for csound Copyright (c) 2000-2014
 * Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB. If not, write to the Free
 * Software Foundation Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307
 * USA
 */

package blue.soundObject;

import blue.*;
import blue.noteProcessor.NoteProcessorChain;
import blue.orchestra.GenericInstrument;
import blue.orchestra.Instrument;
import blue.utility.SoundFileUtilities;
import electric.xml.Element;
import java.io.IOException;
import java.io.Serializable;
import java.util.Map;
import javax.swing.JOptionPane;

/**
 * @author Steven Yi
 *
 */
public class AudioFile extends AbstractSoundObject implements Serializable,
        Cloneable {

    private String soundFileName;

    private String csoundPostCode;

    private boolean useCustomWindowSize = false;

    private int windowSize = 8;

    public AudioFile() {
        this.setName("Audio File");

        soundFileName = "";
        csoundPostCode = "\touts\taChannel1, aChannel1";

    }

    // TODO - EXCEPTION - Look at Code to determine if Exception is Needed
    public NoteList generateNotes(int instrumentNumber, float renderStart, 
            float renderEnd) throws SoundObjectException {
        NoteList n = new NoteList();

        float newDur = subjectiveDuration;

        if(renderEnd > 0 && renderEnd < subjectiveDuration) {
            newDur = renderEnd;
        }

        newDur = newDur - renderStart;

        StringBuilder buffer = new StringBuilder();

        buffer.append("i").append(instrumentNumber);
        buffer.append("\t").append(startTime + renderStart);
        buffer.append("\t").append(newDur);
        buffer.append("\t").append(renderStart);

        Note tempNote = null;

        try {
            tempNote = Note.createNote(buffer.toString());
        } catch (NoteParseException e) {
            throw new SoundObjectException(this, e);
        }
        if (tempNote != null) {
            n.addNote(tempNote);
        }

        return n;
    }

    public Instrument generateInstrument() {
        
        String instrumentText = generateInstrumentText();
        
        if (instrumentText == null) {
            return null;
        }

        GenericInstrument temp = new GenericInstrument();
        temp.setName(this.name);
        temp.setText(instrumentText);
        temp.setEnabled(true);
        //int iNum = arrangement.addInstrument(temp);
        return temp;
    }

    private String generateInstrumentText() {
        StringBuilder iText = new StringBuilder();
        String channelVariables = getChannelVariables();

        if (channelVariables == null) {
            return null;
        }

        String sfName = getSoundFileName().replace('\\', '/');

        iText.append(channelVariables).append("\tdiskin2\t\"");
        iText.append(sfName);
        iText.append("\", 1, p4\n");
        iText.append(getCsoundPostCode());

        return iText.toString();
    }

    private String getChannelVariables() {

        int numChannels;

        try {
            numChannels = SoundFileUtilities
                    .getNumberOfChannels(getSoundFileName());
        } catch (IOException ioe) {
            JOptionPane.showMessageDialog(null, BlueSystem
                    .getString("soundfile.infoPanel.error.couldNotOpenFile")
                    + " " + getSoundFileName());
            return null;
        } catch (javax.sound.sampled.UnsupportedAudioFileException uae) {
            JOptionPane.showMessageDialog(null, BlueSystem
                    .getString("soundfile.infoPanel.error.unsupportedAudio")
                    + " " + uae.getLocalizedMessage());
            return null;
        }

        if (numChannels <= 0) {
            return null;
        }

        String info = "aChannel1";

        int i = 1;

        while (i < numChannels) {
            i++;
            info += ", aChannel" + i;
        }

        return info;
    }

    public int findPowerOfTwo(float seconds) {
        int sr = 44100;
        int samples = Math.round(seconds * sr);

        int powTwoSamples = 2;

        while (powTwoSamples < samples) {
            powTwoSamples = powTwoSamples * 2;
        }

        return powTwoSamples;
    }

    public float getObjectiveDuration() {
        return subjectiveDuration;
    }

    public NoteProcessorChain getNoteProcessorChain() {
        return null;
    }

    public void setNoteProcessorChain(NoteProcessorChain chain) {
    }

    public int getTimeBehavior() {
        return SoundObject.TIME_BEHAVIOR_NOT_SUPPORTED;
    }

    public void setTimeBehavior(int timeBehavior) {
    }

    public float getRepeatPoint() {
        return -1.0f;
    }

    public void setRepeatPoint(float repeatPoint) {
    }


    // METHODS SPECIFIC FOR THIS SOUNDOBJECT

    public String getCsoundPostCode() {
        return csoundPostCode;
    }

    public void setCsoundPostCode(String string) {
        csoundPostCode = string;
    }

    public String getSoundFileName() {
        return soundFileName;
    }

    public void setSoundFileName(String string) {
        soundFileName = string;
    }

    /*
     * (non-Javadoc)
     *
     * @see blue.soundObject.SoundObject#loadFromXML(electric.xml.Element)
     */
    public static SoundObject loadFromXML(Element data,
            Map<String, Object> objRefMap) throws Exception {
        AudioFile aFile = new AudioFile();

        SoundObjectUtilities.initBasicFromXML(data, aFile);

        String sFileName = data.getElement("soundFileName")
                        .getTextString();

        if(sFileName != null) {
            aFile.setSoundFileName(sFileName);
        }
        
        aFile.setCsoundPostCode(data.getElement("csoundPostCode")
                .getTextString());

        return aFile;

    }

    /*
     * (non-Javadoc)
     *
     * @see blue.soundObject.SoundObject#saveAsXML()
     */
    public Element saveAsXML(Map<Object, String> objRefMap) {
        Element retVal = SoundObjectUtilities.getBasicXML(this);

        retVal.addElement("soundFileName").setText(this.getSoundFileName());
        retVal.addElement("csoundPostCode").setText(this.getCsoundPostCode());

        return retVal;
    }

    @Override
    public NoteList generateForCSD(CompileData compileData, float startTime, 
            float endTime)  throws SoundObjectException {
        Instrument instr = this.generateInstrument();
        if(instr == null) {
             throw new RuntimeException(new SoundObjectException(this, BlueSystem
                    .getString("audioFile.couldNotGenerate")
                    + " " + getSoundFileName()));
        }
        int instrNum = compileData.addInstrument(instr);
        NoteList nl = this.generateNotes(instrNum, startTime, endTime);
        return nl;
    }

}

File: packages/blue-data/src/sound-objects/audio-file.ts
/**
 * AudioFile — a SoundObject that plays an audio file via diskin2.
 * Mirrors the Java AudioFile class.
 *
 * Generates a diskin2-based GenericInstrument added to the arrangement
 * and a single i-statement note with skip time as p4.
 */
import { AbstractSoundObject } from './abstract-sound-object';
import { NoteList } from './note-list';
import { Note } from './note';
import { TimeContext } from '../time/time-context';
import { CompileData } from '../compile-data';
import { Element } from '../serialization/xml-reader';
import { ObjRefSaveMap, ObjRefLoadMap } from '../serialization/obj-ref-map';
import { SoundObject } from './sound-object';
import { TimeBehavior } from './time-behavior';
import { initBasicFromXML, getBasicXML } from './sound-object-utilities';
import { GenericInstrument } from '../instruments/generic-instrument';
import type { ScoreGenerationOptions } from '../score/score-generation-options';

export class AudioFile extends AbstractSoundObject {
  private _soundFileName = '';
  private _csoundPostCode = '\touts\taChannel1, aChannel1';
  private _useCustomWindowSize = false;
  private _windowSize = 8;

  constructor(other?: AudioFile) {
    super();
    this.setName('Audio File');
    this._soundFileName = '';
    this._csoundPostCode = '\touts\taChannel1, aChannel1';
    this._useCustomWindowSize = false;
    this._windowSize = 8;
    this._backgroundColor = 0x404040;
    if (other) {
      this.copyFrom(other);
      this._soundFileName = other._soundFileName;
      this._csoundPostCode = other._csoundPostCode;
      this._useCustomWindowSize = other._useCustomWindowSize;
      this._windowSize = other._windowSize;
    }
  }

  getSoundFileName(): string { return this._soundFileName; }
  setSoundFileName(name: string): void { this._soundFileName = name; }

  getCsoundPostCode(): string { return this._csoundPostCode; }
  setCsoundPostCode(code: string): void { this._csoundPostCode = code; }

  useCustomWindowSize(): boolean { return this._useCustomWindowSize; }
  setUseCustomWindowSize(val: boolean): void { this._useCustomWindowSize = val; }

  getWindowSize(): number { return this._windowSize; }
  setWindowSize(size: number): void { this._windowSize = size; }

  override getTimeBehavior(): TimeBehavior {
    return TimeBehavior.NOT_SUPPORTED;
  }

  override generateForCSD(
    context: TimeContext,
    compileData: CompileData,
    startTime: number,
    endTime: number,
    options?: ScoreGenerationOptions,
  ): NoteList {
    if (!this._soundFileName) {
      return new NoteList();
    }

    const instr = this.generateInstrument();
    if (!instr) {
      return new NoteList();
    }

    const instrumentNumber = compileData.addInstrument(instr);
    if (options?.trackId) {
      compileData.addInstrSourceId(instr, options.trackId);
    }

    return this.generateNotes(context, instrumentNumber, startTime, endTime);
  }

  generateInstrument(): GenericInstrument | null {
    const instrText = this.generateInstrumentText();
    if (!instrText) {
      return null;
    }
    const temp = new GenericInstrument();
    temp.setName(this.getName());
    temp.setText(instrText);
    return temp;
  }

  private generateInstrumentText(): string {
    const channelVars = this.getChannelVariables();
    const sfName = this._soundFileName.replace(/\\/g, '/').replace(/"/g, '\\"');
    return `${channelVars}\tdiskin2\t"${sfName}", 1, p4\n${this._csoundPostCode}\n`;
  }

  private getChannelVariables(): string {
    const matches = [...this._csoundPostCode.matchAll(/aChannel(\d+)/g)];
    const indices = matches.map((m) => parseInt(m[1], 10)).filter((n) => !isNaN(n) && n > 0);
    const maxChan = indices.length > 0 ? Math.max(1, ...indices) : 1;
    return Array.from({ length: maxChan }, (_, i) => `aChannel${i + 1}`).join(', ');
  }

  private generateNotes(
    context: TimeContext,
    instrumentNumber: number,
    renderStart: number,
    renderEnd: number,
  ): NoteList {
    const notes = new NoteList();
    const subjectiveDuration = this._subjectiveDuration.toBeats(context);
    let newDur = subjectiveDuration;

    if (renderEnd > 0 && renderEnd < subjectiveDuration) {
      newDur = renderEnd;
    }
    newDur -= renderStart;

    if (!Number.isFinite(newDur) || newDur <= 0) {
      return notes;
    }

    const note = new Note();
    note.setPField(String(instrumentNumber), 1);
    note.setStartTime(this._startTime.toBeats(context) + renderStart);
    note.setSubjectiveDuration(newDur);
    note.setPField(String(renderStart), 4);
    notes.add(note);

    return notes;
  }

  override saveAsXML(_objRefMap?: ObjRefSaveMap): Element {
    const elem = getBasicXML(this, 'blue.soundObject.AudioFile');
    elem.addElement('soundFileName').setText(this._soundFileName);
    elem.addElement('csoundPostCode').setText(this._csoundPostCode);
    return elem;
  }

  static loadFromXML(data: Element, _objRefMap?: ObjRefLoadMap): AudioFile {
    const obj = new AudioFile();
    initBasicFromXML(obj, data);

    const sf = data.getTextString('soundFileName');
    if (sf !== null) obj._soundFileName = sf;

    const post = data.getTextString('csoundPostCode');
    if (post !== null) obj._csoundPostCode = post;

    return obj;
  }

  override deepCopy(): SoundObject {
    return new AudioFile(this);
  }
}
